### ISSUE TITLE
Warning when version range upper version does not end in `-0`

### ISSUE THREAD

**Issue by zivkan** [NuGet/Home#12423]
### NuGet Product(s) Involved

NuGet.exe, MSBuild.exe, dotnet.exe

### The Elevator Pitch

According to SemVer2, 1.0.0 < 2.0.0-alpha < 2.0.0.  But customers frequently use range `[1.2.3, 2.0.0)` hoping to limit the range to `1.x`, but this doesn't take into account pre-release versions.

Therefore, I propose a warning when:
* it finds a package where an upper version is specified
* the upper version is not inclusive (`)` rather than `]`)
* the upper version does not have a prerelease label, then a warning is shown.

Something similar to:

> Warning NUxxxx: Package Contoso.Utilities requests versions [1.2.3, 2.0.0) but will match pre-release version such as 2.0.0-alpha. To exclude prerelease versions, change the version range to [1.2.3, 2.0.0-0)

I'm not sure if this should only be a pack warning (dependency version), only be a restore warning (PackageReference version), or both.

### Additional Context and Details

Consider issues like:
* https://github.com/NuGet/Home/issues/6434
* https://github.com/NuGet/Home/issues/12082

I agree that the behaviour is non-intuitive, and that many customers would prefer the version range comparison to exclude the pre-release versions instead, but it's a breaking change whose magnitude of impact is unknown. This proposal could help package authors in particular minimize risk that their packages claim compatibility with pre-release packages.

**Comment by DavidBoike** [comment]
This warning seems like it would generate a ton of confusion amongst the thousands of developers that would have no idea what was going on. Maybe fixing https://github.com/NuGet/Home/issues/6434 would actually be more straightforward. :wink:

**Comment by zivkan** [comment]
Is the proposed warning message not sufficiently clear?  I know that working on NuGet I know the SemVer spec better than the average developer, but I tried to make the message clear enough to reduce confusion, not add to it.

**Comment by ramonsmits** [comment]
That would be super strange. Don't forget that now suddenly you see such version range on Nuget.org listed everywhere.

This change doesn't make sense IMHO. Also don't forget there are thousands of packages without the `-0` workaround that expect that behavior to not match prereleases for the right / max side of the range.

This change wouldn't fix any of these already created packages. Fixing the version matching would actually fix incorrect behavior for all those thousands of packages without needed to push a new "patch" release.

**Comment by DavidBoike** [comment]
> Is the proposed warning message not sufficiently clear? I know that working on NuGet I know the SemVer spec better than the average developer, but I tried to make the message clear enough to reduce confusion, not add to it.

No, I was being sarcastic. Sorry about that. I'm saying the warning should not exist, because imagine what all those developers who think they know how NuGet works start getting this warning telling them they need to do this odd workaround that only must exist because nobody will fix https://github.com/NuGet/Home/issues/6434. As much discussion in that issue is complaining about how fixing it would be a breaking change, it seems like that big breaking change would probably be _less problematic_ than a warning like this.

I like this issue from the angle of "If you're not going to fix it, then you really should do this instead" to illustrate that the "breaking change" is the better course of action.

But it's worth pointing out that in https://github.com/NuGet/Home/issues/6434 there's plenty of evidence that the `-0` suffix doesn't really work appropriately either.

**Comment by zivkan** [comment]
I missed the comment that said NU5104 was being raised when an exclusive upper bound uses `-0`. I think that's a bug in the pack warning, so I created this issue:
* https://github.com/NuGet/Home/issues/12424

**Comment by nkolev92** [comment]
When a range is specified with either a lower or upper boundary of prerelease it opts in the package into prerelease versions, so I'm not certain it's the right behavior. By specifying a stable range only, you'll only get prerelease versions if another component in the graph asks for a prerelease version of the package.

When only stable versions are specified by the project or dependencies, then only stable versions will be considered. That is how the https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1103 warnings comes to be, which warns about not finding a matching stable version of a package while a prerelease is available.

NuGet *will not* choose a prerelease version for a specific package unless someone asks for a prerelease versions which adding an upper range with a prerelease effectively does.

**Comment by DavidBoike** [comment]
But that doesn't make sense either.

Why should the fact that the [current version of the Lucene.Net package is 4.8.0-beta00016](https://www.nuget.org/packages/Lucene.Net/4.8.0-beta00016) and that it has _been in beta since May 2017_ make a project that wants the latest NServiceBus 7.x patch release eligible for [NServiceBus 8.0.0-beta.2](https://www.nuget.org/packages/NServiceBus/8.0.0-beta.2) that contains binary breaking changes?

**Comment by nkolev92** [comment]
@DavidBoike 

Another clarification, the opt-in into prerelease works on a per id basis. 

Package A requesting a prerelease version does not affect the selection for package B if only stable versions are being requested. 

The simplest example: 

Project specifies: 

A 1.0.0
B 1.0.0-beta.1

Source has:
A 1.1.0-beta.1
A 1.1.0

B 1.1.0-beta.1
B 2.0.0

A 1.1.0 will be resolved because a stable version is requested despite the prerelease version being lower.
B 1.1.0-beta.1 will be resolved as prerelease versions for B are allowed, and it's closer than 2.0.0.

**Comment by Frulfump** [comment]
> This issue will now be closed since it had been marked Icebox cleanup candidate, but received no further activity in the past 14 days.

Closing as completed is definitely wrong, should be closed as not planned if any type of closure is necessary due to inactivity.

This still seems reasonable, I wasn't aware of needing to use -0 at least to get it to ignore pre-release versions. It's not mentioned here https://learn.microsoft.com/en-us/nuget/concepts/package-versioning?tabs=semver20sort#version-ranges which it probaby should?

**Comment by ramonsmits** [comment]
Clearly seems this will never get fixed and everybody is left to use the `-0` workaround that is very unnatural like `[1.2.3, 2.0.0-0)`.

If at mininum indeed those docs would mention that and this quirk is integrated in the XML and tables as all those examples pretty sure nobody intentionally would want to include a pre-release version of ANY kind unless you explicitly wanted to.

Notation	|Applied rule|	Description
----|---|---
(,1.0-0]	|x ≤ 1.0	|Maximum version, inclusive
(,1.0-0)	|x < 1.0	|Maximum version, exclusive
[1.0,2.0-0]	|1.0 ≤ x ≤ 2.0	|Exact range, inclusive
(1.0,2.0-0)	|1.0 < x < 2.0	|Exact range, exclusive
[1.0,2.0-0)	|1.0 ≤ x < 2.0	|Mixed inclusive minimum and exclusive maximum version

**Comment by nkolev92** [comment]
Hey @Frulfump, @ramonsmits 

I want to bring attention to my comment above.

> Another clarification, the opt-in into prerelease works on a per id basis.

As long as all anyone is dealing with is stable versions, then the concern is not there.

So even if prereleases are available, as long as the specification itself isn't explicitly allowing prerelease versions, it won't pick them.

**Comment by ramonsmits** [comment]
> So even if prereleases are available, as long as the specification itself isn't explicitly allowing prerelease versions, it won't pick them.

@nkolev92 that pretty much is one of the big issues.

When a maintainer sets a version based on semver like [2,3) on a dependency it expresses that it only knows it is compatible with release and pre-release packages. That should simply not match ANY 3.0.0 pre-release package, while it currently does and that does not make any sense.

But we are repeating... it seems the team simply does not want to confirm the friction and resolve it so last comment here
